source/device_defender.c

```diff
@@ -316,318 +565,832 @@ static uint64_t s_defender_report_id_epoch_time_ms(struct aws_iotdevice_defender
         uint64_t now;
-        aws_event_loop_current_clock_time(defender_task->config.event_loop, &now);
+        aws_event_loop_current_clock_time(defender_task->event_loop, &now);
         aws_event_loop_schedule_task_future(
-            defender_task->config.event_loop, task, now + defender_task->config.task_period_ns);
+            defender_task->event_loop, task, now + defender_task->config.task_period_ns);
     }
 
-    goto cleanup;
-
 cleanup:
+    s_clean_up_custom_metric_data(allocator, custom_metric_data, custom_metrics_len);
 
     if (aws_hash_table_is_valid(&ifconfig.iface_name_to_info)) {
         aws_hash_table_clean_up(&ifconfig.iface_name_to_info);
     }
 }
 
-struct aws_iotdevice_defender_v1_task *aws_iotdevice_defender_v1_report_task(
+int aws_iotdevice_defender_config_create(
+    struct aws_iotdevice_defender_task_config **config_out,
     struct aws_allocator *allocator,
-    const struct aws_iotdevice_defender_report_task_config *config) {
+    const struct aws_byte_cursor *thing_name,
+    enum aws_iotdevice_defender_report_format report_format) {
+    AWS_PRECONDITION(config_out != NULL);
     AWS_PRECONDITION(aws_allocator_is_valid(allocator));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(thing_name));
+
+    struct aws_iotdevice_defender_task_config *config = NULL;
+
+    if (report_format != AWS_IDDRF_JSON) {
+        aws_raise_error(AWS_ERROR_IOTDEVICE_DEFENDER_UNSUPPORTED_REPORT_FORMAT);
+        goto error_return;
+    }
+
+    config = aws_mem_acquire(allocator, sizeof(struct aws_iotdevice_defender_task_config));
+    AWS_ZERO_STRUCT(*config);
+
+    config->thing_name = aws_string_new_from_cursor(allocator, thing_name);
+    config->allocator = allocator;
+    config->report_format = report_format;
+    config->callback_userdata = NULL;
+    config->task_canceled_fn = NULL;
+    config->rejected_report_fn = NULL;
+    config->accepted_report_fn = NULL;
+    /* defaults here will be consistent across any language on top */
+    config->task_period_ns = 5ULL * 60ULL * 1000000000ULL;
+    aws_array_list_init_dynamic(&config->custom_metrics, allocator, 0, sizeof(struct defender_custom_metric *));
+    config->custom_metrics_len = 0;
+
+    *config_out = config;
+    return AWS_OP_SUCCESS;
+
+error_return:
+    aws_mem_release(allocator, config);
+
+    return AWS_OP_ERR;
+}
+
+void aws_iotdevice_defender_config_clean_up(struct aws_iotdevice_defender_task_config *config) {
     AWS_PRECONDITION(config != NULL);
-    bool failure = false;
-    struct aws_iotdevice_defender_v1_task *defender_task = NULL;
+    struct aws_allocator *allocator = config->allocator;
+    AWS_ASSERT(aws_allocator_is_valid(allocator));
+    /* assign caller ptr to NULL */
+    if (config) {
+        s_defender_config_clean_up_internals(config);
+        aws_mem_release(allocator, config);
+    }
+}
+
+/* Maybe highlights the need for aws_string construction to consume a aws_byte_buf and own it?? */
```

Comment:
We don't leave TODO comments. 

To your point though, a string and a byte_buf are separate resources right, so you'd expect to have their init/new and clean_up/destroy functions called. You could certainly write some utility functions to transfer "ownership" of a buffer, but because a string is dynamically allocated and has its length at its head (in terms of memory layout), you can't actually re-use the memory.